File: python/ldpc_decoder.py

class ldpc_decoder:
    
    """ LDPC Decoder 
    
    Initialize with the desired alist
    
    See http://www.inference.org.uk/mackay/codes/alist.html for
        documentation on the .alist format
        
    alist -- string, name of alist file
    
    """
    
    def __init__(self, alist):
        
        self.checknodes, self.N, self.M = self.read_checknodes_from_alist(alist)
        
    def read_checknodes_from_alist(self, alist):
   
        """ Read check nodes from an alist file
        
        alist -- string, name of alist file to read
        
       
        returns:
        
        mlist -- list, list of checknodes
        N -- int, N parameter from alist
        M -- int, M parameter from alist
        
        """
        f = open(alist, 'r')
        
        line = f.readline()
        N, M = line.split(' ')
        
        line = f.readline()
        max_col_weight, max_row_weight = line.split(' ')
        
        line = f.readline()
        col_weights = line.split(' ')
        col_weights.pop()
        
        line = f.readline()
        row_weights = line.split(' ')
        row_weights.pop()
        
        nlist = []
        mlist = []
        
        for i in range(int(N)):
            nlist.append(f.readline().split('\n')[0])
        
        for i in range(int(M)):
            mlist.append(f.readline().split('\n')[0].split(' '))
        
        return mlist, int(N), int(M)
        

    def compute_syndrome(self, checknodes = self.checknodes, rx_codeword):
        """Compute the Syndrome
        
        checknodes -- list, checknodes for the given LDPC code
        rx_codeword -- numpy array (uint8), received LDPC codeword
        
        returns:
        
        num_errors -- int, number of errors 
        """
        num_errors = 0
        
        for i in checknodes:
            indices = i
            xorsum = 0
            
            for k in indices:
                if (k != '' and int(k) != 0):
                    xorsum ^= rx_codeword[int(k)-1]
            
            if (xorsum != 0):
                num_errors += 1
        
        return num_errors
        
    def ldpc_tdmp(self, codeword, checknodes = self.checknodes, max_iter):
        
        """Correct Errors using the Turbo-decoding Message-Passing
        (TDMP) algorithm
        
        codeword -- numpy array (float), received codeword
        checknodes -- list, checknodes for LDPC code
        max_iter -- int, max number of TDMP iterations to run
        
        returns:
        result -- numpy array (uint8), resulting codeword after LDPC decoding
        
        """
        
        LMN = np.zeros((max_iter, len(checknodes), len(codeword)))

        result = np.zeros(len(codeword), dtype=np.uint8)
        float_in = np.zeros(len(codeword))
        
        for n in range(len(codeword)):
            if codeword[n] == 0:
                float_in[n] = 1.0
            else:
                float_in[n] = -1.0
        
        

        for i in range(max_iter):
            
            # loop through all checknodes
            for m in range(len(checknodes)):
                
                nodes = checknodes[m]    

                nodes = [ int(x)-1 for x in nodes if x != '0' and x != '']

                LNM = np.zeros(len(nodes))
                
                # compute LNM message
                if i == 0:
                    for n in range(len(nodes)):
                        LNM[n] = float_in[nodes[n]]
                else:
                    for n in range(len(nodes)):
                        LNM[n] = float_in[nodes[n]] - LMN[i-1][m][nodes[n]]
                        
                # compute LMN message        
                for n in range(len(nodes)):
                    node = nodes[n]
                    
                    sign = 1.0
                    minimum = 10000.0
                    
                    for d in range(len(nodes)):
                        if d != n:
                            sign *= LNM[d]
                            
                            if abs(LNM[d])< minimum:
                                minimum = abs(LNM[d])
                                
                    if sign > 0:
                        LMN[i][m][node] =  minimum
                    else:
                        LMN[i][m][node] =  -minimum
                        
                # update codeword        
                for n in range(len(nodes)):
                    float_in[nodes[n]] = LNM[n] + LMN[i][m][nodes[n]]
                    
        # hard decision    
        for n in range(len(codeword)):
            if float_in[n] <= 0:
                result[n] = 1
            else:
                result[n] = 0
                
        return result        

File: lib/ldpc_decoder.cc
#include <stdio.h>
#include <cmath>
#include <cstdlib>
#include <cstring>

#include <volk/volk.h>
#include <volk/volk_malloc.h>

#include "ldpc_decoder.h"

//#define PRINT_ERRORS

// Right now the flags and values 
// for the min sum algorithm modifications
// are here.  These values were found experimentally.
// Only select one at a time or comment both out.
#define NMS
#define NMS_OFFSET 0.815

//#define OMS
#define OMS_OFFSET .10

namespace wimax_ldpc_lib{
    
    ldpc_decoder::ldpc_decoder() {}
    
    ldpc_decoder::ldpc_decoder(coderate rate,
                            unsigned int z_factor,
                            unsigned int max_iter,
                            unsigned int num_threads) : ldpc(rate, z_factor, num_threads)
    {
        m_max_iter = max_iter;

        m_alignment = volk_get_alignment();

        m_LMN = (float*)volk_malloc(sizeof(float) * (m_max_iter+1) * m_row_size * m_col_size, m_alignment);

        std::memset(m_LMN, 0, sizeof(float) * m_max_iter * m_row_size * m_col_size); 

        m_row_lens = (unsigned int*)malloc(sizeof(int) * m_row_size);

        unsigned int current_row_len = 0;

        // measure the length of each row in the checknode array
        // doing this ahead of time is better than doing it during 
        // decoding.
        for (unsigned int m = 0; m < m_row_size; m++)
        {
            current_row_len = 0;
            for (unsigned int n = 0; n < m_col_size; n++)
            {
                if (m_checknode_array[m*m_col_size + n] != -1)
                {
                    current_row_len += 1;
                }
            }
            m_row_lens[m] = current_row_len;
        }
    }

    ldpc_decoder::~ldpc_decoder()
    {
        volk_free(m_LMN);
        free(m_row_lens);
    }

    unsigned int ldpc_decoder::compute_syndrome(int8_t* rx_codeword, bool early_exit)
    {
        unsigned int num_errors = 0;

        for (unsigned int i = 0; i < m_row_size; i++)
        {
            int8_t xorsum = 0;
            int16_t temp_index;
            unsigned int offset = i*m_col_size;

            for (unsigned int k = 0; k < m_col_size; k++)
            {
                temp_index = m_checknode_array[offset + k];
                if (temp_index != -1)
                {
                    xorsum ^= rx_codeword[temp_index];
                }
            }

            // if we just want to know if there are errors or
            // not, then return when the first error is found.
            // This saves time during decoding
            if (xorsum != 0)
            {
                if (early_exit)
                {
                    num_errors = 1;
                    i = m_row_size;
                }
                else
                {
                    num_errors += 1;
                }
            }
        }

        return num_errors;
    }

    // this is function takes a bit array and converts it
    // to the input required by the decoder.
    unsigned int ldpc_decoder::decode(int8_t* rx_codeword, int8_t* decoded)
    {
        float soft_codeword[m_N] __attribute__((aligned(32)));

        // convert to soft demod
        for(unsigned int i = 0; i < m_N; i++)
        {
            soft_codeword[i] = rx_codeword[i] * 2.0 - 1.0;
        }
        return this->decode(soft_codeword, decoded);
    }

    // TDMP Decoder for LDPC
    // takes soft demodulated input
    unsigned int ldpc_decoder::decode(float* rx_codeword, int8_t* decoded)
    {
        unsigned int num_errors = 0;
        unsigned int initial_errors = 0;

        // Check for errors initially, if there are none
        // then exit.
        // Perform hard decision decode
        volk_32f_binary_slicer_8i(decoded, rx_codeword, m_N);

        // compute the symdrome
        initial_errors = this->compute_syndrome(decoded, true);

        // Exit if there are no inital errors
        if(initial_errors == 0)
        {
            return initial_errors;
        }

        for (unsigned int i = 1; i < m_max_iter+1; i++)
        {
            for (unsigned int m = 0; m < m_row_size; m++)
            {
                // declaring this local to the loop helps
                float LNM[m_row_lens[m]];

                bool sign = true;

                float first_minimum = 10000.0;
                float minimum;
                float iter_sign;
                float lnm_abs;

                unsigned int cn_offset = m*m_col_size;
                unsigned int lmn_offset;

                unsigned int first_min_index = 0;

                // Compute LNM Message

                lmn_offset = m_col_size*(m + (i-1)*m_row_size);

                for (unsigned int n = 0; n < m_row_lens[m]; n++)
                {

                    LNM[n] = rx_codeword[m_checknode_array[cn_offset + n]] 
                            - m_LMN[n + lmn_offset];

                    // calculate the sign bit here so that
                    // the operation in the next loop is simpler
                    sign ^= !std::signbit(LNM[n]);

                    // calculate the first minimum here so that
                    // we can calculate the minimum fewer times
                    lnm_abs = fabs(LNM[n]);
                    if (lnm_abs < first_minimum)
                    {
                        first_minimum = lnm_abs;
                        first_min_index = n;
                    }
                }

                lmn_offset = m_col_size*(m + i*m_row_size);

                // compute and apply LNM message for starting indices

                for (unsigned int n = 0; n < m_row_lens[m]; n++)
                {

                    iter_sign = (sign ^ !std::signbit(LNM[n])) * -2 + 1;

                    // because we find the first min in the loop above,
                    // this loop only runs one time per iteration.
                    if (n == first_min_index)
                    {
                        // Compute and apply LNM message for min index
                        minimum = 10000.0;

                        for (unsigned int d = 0; d < m_row_lens[m]; d++)
                        {
                            if (d != first_min_index)
                            {
                                lnm_abs = fabs(LNM[d]);
                                if (lnm_abs < minimum)
                                    minimum = lnm_abs;
                            }
                        }
                    }
                    else
                    {
                        minimum = first_minimum;
                    }

                    // The option to perform NMS (Normalized Min Sum)
                    // OMS(Offset Min Sum) or just Min Sum is given below.
                    #if defined(NMS) || defined(OMS)
                    #ifdef NMS

                    m_LMN[n + lmn_offset] = iter_sign * (minimum*NMS_OFFSET);

                    rx_codeword[m_checknode_array[cn_offset + n]] = 
                        LNM[n] + iter_sign * (minimum*NMS_OFFSET);

                    #else

                    m_LMN[n + lmn_offset] = iter_sign * std::max(minimum-OMS_OFFSET, 0.0);

                    rx_codeword[m_checknode_array[cn_offset + n]] = 
                        LNM[n] + iter_sign *
                        std::max(minimum-OMS_OFFSET, 0.0);
                    #endif

                    #else

                    m_LMN[n + lmn_offset] = iter_sign * (minimum);
                    rx_codeword[m_checknode_array[cn_offset + n]] = LNM[n] + iter_sign * (minimum);
                    #endif
                }
            }

            volk_32f_binary_slicer_8i(decoded, rx_codeword, m_N);

            // compute the symdrome
            num_errors = this->compute_syndrome(decoded, true);

            // Exit if there are no more errors
            if(num_errors == 0)
            {
                return num_errors;
            }
        }

        #ifdef PRINT_ERRORS
        printf("Initial Errors: %d\n", initial_errors);
        printf("Final Errors: %d\n", num_errors);
        #endif

        // return the number of errors remaining after 
        // the maximum number of iterations is performed.
        return num_errors;
    }
} // end namespace wimax_ldpc_lib
